### ISSUE TITLE
Question about DiffLoss

### ISSUE THREAD

**Issue by kmaninis** [fungtion/DSN#2]
Hi, thanks for the nice PyTorch implementation.

I have some questions for the DiffLoss:
- Shouldn't [this](https://github.com/fungtion/DSN/blob/e09117ba961dacd1f1cea7657a5f70a9df4fa51b/functions.py#L63) line compute the correlation of each feature dimension of the private features to each feature dimension of the shared features? As it is now, it is computing the correlation of one sample to another. Correct me if I'm wrong, but shouldn't it be:
`torch.mean((input1_l2.t().mm(input2_l2).pow(2)))` instead?

- Also, you mention that there are some stability issues. Could that be because there is no mean value normalization, as done in the [TF implementation](https://github.com/tensorflow/models/blob/630c4ca89c714bdfb9a16df34913dd9f4c38350a/research/domain_adaptation/domain_separation/losses.py#L205) of the authors?

Thanks a lot :)

**Comment by fungtion** [comment]
@kmaninis Thank you for your feedback. 

- As far as the first question, each row of input1 and input2 in DiffLoss represents the private and shared feature of the same sample, respectively. The DiffLoss is supposed to make the private and shared features orthogonal. So I think it should be `torch.mean((input1_l2.mm(input2_l2.t()).pow(2)))`.
- For the second question, I don't think mean value normalization could improve the stability significantly, since the the proposed DiffLoss has brought too much instability.

**Comment by kmaninis** [comment]
@fungtion Thanks for your reply.

I think that the orthogonality constraint in your case doesn't hold in case you permute your features, for example [0, 1] and [1, 0] are orthogonal, but you can get the one from the other by just permuting the features.

In short, I think that what it is meant to do is: For a private feature of dimension M x C1 and a shared representation M x C2 (M stands for batch size, C stands for features) create a correlation C1 X C2 and minimize it. Isn't it what is happening in [this](https://github.com/tensorflow/models/blob/8ff6115343e37f6d62cea3f4ab26127799bf8775/research/domain_adaptation/domain_separation/losses.py#L211) part of the code?

**Comment by fungtion** [comment]
@kmaninis Yes, I see what you mean, this is probably my misunderstanding about the paper, which makes it unstable. I'll try to modify my implementation. Thanks !
